lib/matplotlib/dates.py

```diff
@@ -740,12 +746,237 @@ def __call__(self, x, pos=0):
+        self._oldlabels = None
+        self.defaultfmt = '%Y'
+        # there are 6 levels with each level getting a specific format
+        # 0: mostly years,  1: months,  2: days,
+        # 3: hours, 4: minutes, 5: seconds
+        if formats:
+            if len(formats) != 6:
+                raise ValueError('formats argument must be a list of '
+                                 '6 format strings (or None)')
+            self.formats = formats
+        else:
+            self.formats = ['%Y',  # ticks are mostly years
+                            '%b',          # ticks are mostly months
+                            '%d',          # ticks are mostly days
+                            '%H:%M',       # hrs
+                            '%H:%M',       # min
+                            '%S.%f',       # secs
+                            ]
+        # fmt for zeros ticks at this level.  These are
+        # ticks that should be labeled w/ info the level above.
+        # like 1 Jan can just be labled "Jan".  02:02:00 can
+        # just be labeled 02:02.
+        if zero_formats:
+            if len(formats) != 6:
+                raise ValueError('zero_formats argument must be a list of '
+                                 '6 format strings (or None)')
+            self.zero_formats = zero_formats
+        elif formats:
+            # use the users formats for the zero tick formats
+            self.zero_formats = [''] + self.formats[:-1]
+        else:
+            # make the defaults a bit nicer:
+            self.zero_formats = [''] + self.formats[:-1]
+            self.zero_formats[3] = '%b-%d'
+
+        if offset_formats:
+            if len(offset_formats) != 6:
+                raise ValueError('offsetfmts argument must be a list of '
+                                 '6 format strings (or None)')
+            self.offset_formats = offset_formats
+        else:
+            self.offset_formats = ['',
+                                   '%Y',
+                                   '%Y-%b',
+                                   '%Y-%b-%d',
+                                   '%Y-%b-%d',
+                                   '%Y-%b-%d %H:%M']
+        self.offset_string = ''
+        self._formatter = DateFormatter(self.defaultfmt, self._tz)
+        self.show_offset = show_offset
+
+    def __call__(self, x, pos=None):
+        if hasattr(self._locator, '_get_unit'):
+            locator_unit_scale = float(self._locator._get_unit())
+        else:
+            locator_unit_scale = 1.0
+        ticks = self._locator()
+        if pos is not None:
+            if not np.array_equal(ticks, self._oldticks):
+
+                offset_fmt = ''
+                fmt = self.defaultfmt
+                self._formatter = DateFormatter(fmt, self._tz)
+                tickdate = np.array([num2date(tick).timetuple()[:6]
+                                     for tick in ticks])
+                tickdatetime = [num2date(tick) for tick in ticks]
```

Comment:
If you flip the order of these two lines, you won't need to call `num2date` twice and create duplicate `datetime` objects, I think?
```suggestion
                tickdatetime = [num2date(tick) for tick in ticks]
                tickdate = np.array([tick.timetuple()[:6]
                                     for tick in tickdatetime]) 
```